**So now we have information and have extracted enough features to move forward for next phase-model training**

In [33]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [34]:
import pandas as pd
data=pd.read_csv('/content/drive/MyDrive/processed_data.csv')


**We will do**

1.  Linear Regression model
2.  Random forest regressor





**1.   Linear Regression**:



In [35]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              61 non-null     object 
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  object 
 6   weather_description  48204 non-null  object 
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
 9   hour                 48204 non-null  int64  
 10  weekdays             48204 non-null  int64  
 11  month                48204 non-null  int64  
 12  year                 48204 non-null  int64  
 13  is_holiday           48204 non-null  int64  
 14  is_weekend           48204 non-null  int64  
 15  is_rush_hour         48204 non-null 

In [36]:
X = data.drop( [    'traffic_volume','holiday','date_time' ,  'weather_main',
                'weather_description' ],   axis=1)

y = data['traffic_volume']


In [37]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [38]:
from sklearn.linear_model import LinearRegression
model=LinearRegression()

model.fit(X_train,y_train)


LinearRegression()

In [39]:
y_pred=model.predict(X_test)

In [40]:
import pandas as pd
comparison = pd.DataFrame({'actual': y_test,'predicted': y_pred})
print(comparison.head())

       actual    predicted
35748    6364  4814.042482
147      2030  3465.433132
24362     708  2095.911250
35495    2029  4082.134308
11010    3734  2636.234277


**Now Evalaution of model**

In [41]:
from sklearn.metrics import ( mean_absolute_error,mean_squared_error,r2_score)

In [42]:
import numpy as np
mae = mean_absolute_error( y_test,y_pred)

rmse = np.sqrt(
mean_squared_error( y_test,y_pred))

r2 = r2_score(y_test,y_pred)


In [43]:
print("MAE :", mae)
print("RMSE:", rmse)
print("R²  :", r2)


MAE : 1312.882279873917
RMSE: 1518.4122202748335
R²  : 0.41682869269946365


**Linear Regression Results**

MAE  : 1314.08
RMSE : 1519.24
R²   : 0.4162

**The model explains approximately 41.6% of the variance in traffic volume. The average prediction error is around 1314 vehicles. These results establish a baseline for comparison with more advanced machine learning models**




**2. Random Forest Regressor**:




In [44]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(n_estimators=100, random_state=42)

In [45]:
rf.fit( X_train,y_train)

RandomForestRegressor(random_state=42)

In [46]:
rf_pred = rf.predict(X_test)

In [47]:
comparison = pd.DataFrame({'Actual': y_test, 'Predicted': rf_pred})
print(comparison.head())

       Actual  Predicted
35748    6364    5396.46
147      2030    2237.22
24362     708     678.86
35495    2029    2065.71
11010    3734    4874.02


In [48]:
mae = mean_absolute_error( y_test,rf_pred)

rmse = np.sqrt(
mean_squared_error( y_test,rf_pred))

r2 = r2_score(y_test, rf_pred)

In [49]:
print("MAE :", mae)
print("RMSE:", rmse)
print("R²  :", r2)

MAE : 208.62724078052895
RMSE: 407.325820872871
R²  : 0.9580337029656465


**Random Forest Results**

MAE  : 208.66
RMSE : 407.32
R²   : 0.9580

**The Random Forest Regressor significantly outperformed the Linear Regression model. The model achieved an R² score of 0.958, indicating that it explains approximately 95.8% of the variance in traffic volume. The substantial reduction in MAE and RMSE demonstrates the model's ability to accurately capture complex and non-linear traffic patterns.**